Suggest removing `&mut x` when `x: &mut T` and `&mut T` is expected type

An example:
```rust
struct A;

fn bar(x: &mut A) {}

fn foo(x: &mut A) {
    bar(&mut x);
}
```
results in:
```rust
error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
 --> src/lib.rs:6:9
  |
5 | fn foo(x: &mut A) {
  |        - help: consider changing this to be mutable: `mut x`
6 |     bar(&mut x);
  |         ^^^^^^ cannot borrow as mutable
```
but we could say:
```rust
error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
 --> src/lib.rs:6:9
  |
5 | fn foo(x: &mut A) {
6 |     bar(&mut x);
  |         ^^^^^^ cannot borrow as mutable
  |         ---- help: remove the unnecessary `&mut` here
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"ibraheemdev"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

I think this issue was fixed:
```rust
error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
 --> src/lib.rs:6:9
  |
6 |     bar(&mut x);
  |         ^^^^^^
  |         |
  |         cannot borrow as mutable
  |         try removing `&mut` here
```

--- a/compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs
+++ b/compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs
@@ -240,7 +240,12 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                     .unwrap_or(false) =>
             {
                 err.span_label(span, format!("cannot {ACT}", ACT = act));
-                err.span_label(span, "try removing `&mut` here");
+                err.span_suggestion(
+                    span,
+                    "try removing `&mut` here",
+                    String::new(),
+                    Applicability::MaybeIncorrect,
+                );
             }
 
             // We want to suggest users use `let mut` for local (user
@@ -322,7 +327,12 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                 } =>
             {
                 err.span_label(span, format!("cannot {ACT}", ACT = act));
-                err.span_label(span, "try removing `&mut` here");
+                err.span_suggestion(
+                    span,
+                    "try removing `&mut` here",
+                    String::new(),
+                    Applicability::MaybeIncorrect,
+                );
             }
 
             PlaceRef { local, projection: [ProjectionElem::Deref] }

--- a/src/test/ui/borrowck/issue-33819.rs
+++ b/src/test/ui/borrowck/issue-33819.rs
@@ -3,7 +3,7 @@ fn main() {
     match op {
         Some(ref v) => { let a = &mut v; },
         //~^ ERROR cannot borrow `v` as mutable, as it is not declared as mutable
-        //~^ HELP try removing `&mut` here 
+        //~^ HELP try removing `&mut` here
         None => {},
     }
 }

--- a/src/test/ui/borrowck/issue-33819.rs
+++ b/src/test/ui/borrowck/issue-33819.rs
@@ -3,6 +3,7 @@ fn main() {
     match op {
         Some(ref v) => { let a = &mut v; },
         //~^ ERROR cannot borrow `v` as mutable, as it is not declared as mutable
+        //~^ HELP try removing `&mut` here 
         None => {},
     }
 }

--- a/src/test/ui/borrowck/issue-33819.rs
+++ b/src/test/ui/borrowck/issue-33819.rs
@@ -3,7 +3,7 @@ fn main() {
     match op {
         Some(ref v) => { let a = &mut v; },
         //~^ ERROR cannot borrow `v` as mutable, as it is not declared as mutable
-        //~^ HELP try removing `&mut` here
+        //~| HELP try removing `&mut` here
         None => {},
     }
 }

--- a/src/test/ui/borrowck/issue-33819.stderr
+++ b/src/test/ui/borrowck/issue-33819.stderr
@@ -5,7 +5,7 @@ LL |         Some(ref v) => { let a = &mut v; },
    |                                  ^^^^^^
    |                                  |
    |                                  cannot borrow as mutable
-   |                                  try removing `&mut` here
+   |                                  help: try removing `&mut` here
 
 error: aborting due to previous error
 

--- a/src/test/ui/borrowck/mut-borrow-of-mut-ref.rs
+++ b/src/test/ui/borrowck/mut-borrow-of-mut-ref.rs
@@ -4,8 +4,10 @@
 pub fn f(b: &mut i32) {
     g(&mut b);
     //~^ ERROR cannot borrow
+    //~| HELP try removing `&mut` here
     g(&mut &mut b);
     //~^ ERROR cannot borrow
+    //~| HELP try removing `&mut` here
 }
 
 pub fn g(_: &mut i32) {}

--- a/src/test/ui/borrowck/mut-borrow-of-mut-ref.stderr
+++ b/src/test/ui/borrowck/mut-borrow-of-mut-ref.stderr
@@ -5,16 +5,16 @@ LL |     g(&mut b);
    |       ^^^^^^
    |       |
    |       cannot borrow as mutable
-   |       try removing `&mut` here
+   |       help: try removing `&mut` here
 
 error[E0596]: cannot borrow `b` as mutable, as it is not declared as mutable
-  --> $DIR/mut-borrow-of-mut-ref.rs:7:12
+  --> $DIR/mut-borrow-of-mut-ref.rs:8:12
    |
 LL |     g(&mut &mut b);
    |            ^^^^^^
    |            |
    |            cannot borrow as mutable
-   |            try removing `&mut` here
+   |            help: try removing `&mut` here
 
 error: aborting due to 2 previous errors
 

--- a/src/test/ui/did_you_mean/issue-31424.rs
+++ b/src/test/ui/did_you_mean/issue-31424.rs
@@ -5,13 +5,16 @@ struct Struct;
 impl Struct {
     fn foo(&mut self) {
         (&mut self).bar(); //~ ERROR cannot borrow
+        //~^ HELP try removing `&mut` here
     }
 
     // In this case we could keep the suggestion, but to distinguish the
     // two cases is pretty hard. It's an obscure case anyway.
     fn bar(self: &mut Self) {
         //~^ WARN function cannot return without recursing
+        //~^^ HELP a `loop` may express intention better if this is on purpose
         (&mut self).bar(); //~ ERROR cannot borrow
+        //~^ HELP try removing `&mut` here
     }
 }
 

--- a/src/test/ui/did_you_mean/issue-31424.stderr
+++ b/src/test/ui/did_you_mean/issue-31424.stderr
@@ -5,14 +5,14 @@ LL |         (&mut self).bar();
    |         ^^^^^^^^^^^
    |         |
    |         cannot borrow as mutable
-   |         try removing `&mut` here
+   |         help: try removing `&mut` here
 
 warning: function cannot return without recursing
-  --> $DIR/issue-31424.rs:12:5
+  --> $DIR/issue-31424.rs:13:5
    |
 LL |     fn bar(self: &mut Self) {
    |     ^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
-LL |
+...
 LL |         (&mut self).bar();
    |         ----------------- recursive call site
    |
@@ -20,13 +20,13 @@ LL |         (&mut self).bar();
    = help: a `loop` may express intention better if this is on purpose
 
 error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
-  --> $DIR/issue-31424.rs:14:9
+  --> $DIR/issue-31424.rs:16:9
    |
 LL |         (&mut self).bar();
    |         ^^^^^^^^^^^
    |         |
    |         cannot borrow as mutable
-   |         try removing `&mut` here
+   |         help: try removing `&mut` here
 
 error: aborting due to 2 previous errors; 1 warning emitted
 

--- a/src/test/ui/did_you_mean/issue-34126.rs
+++ b/src/test/ui/did_you_mean/issue-34126.rs
@@ -5,6 +5,7 @@ impl Z {
     fn start(&mut self) {
         self.run(&mut self); //~ ERROR cannot borrow
         //~| ERROR cannot borrow
+        //~| HELP try removing `&mut` here
     }
 }
 

--- a/src/test/ui/did_you_mean/issue-34126.stderr
+++ b/src/test/ui/did_you_mean/issue-34126.stderr
@@ -5,7 +5,7 @@ LL |         self.run(&mut self);
    |                  ^^^^^^^^^
    |                  |
    |                  cannot borrow as mutable
-   |                  try removing `&mut` here
+   |                  help: try removing `&mut` here
 
 error[E0502]: cannot borrow `self` as mutable because it is also borrowed as immutable
   --> $DIR/issue-34126.rs:6:18

--- a/src/test/ui/did_you_mean/issue-34337.rs
+++ b/src/test/ui/did_you_mean/issue-34337.rs
@@ -4,4 +4,5 @@ fn main() {
     let mut v: Vec<String> = Vec::new();
     let ref mut key = v[0];
     get(&mut key); //~ ERROR cannot borrow
+    //~| HELP try removing `&mut` here
 }

--- a/src/test/ui/did_you_mean/issue-34337.stderr
+++ b/src/test/ui/did_you_mean/issue-34337.stderr
@@ -5,7 +5,7 @@ LL |     get(&mut key);
    |         ^^^^^^^^
    |         |
    |         cannot borrow as mutable
-   |         try removing `&mut` here
+   |         help: try removing `&mut` here
 
 error: aborting due to previous error
 

--- a/src/test/ui/did_you_mean/issue-37139.rs
+++ b/src/test/ui/did_you_mean/issue-37139.rs
@@ -10,6 +10,7 @@ fn main() {
     match x {
         TestEnum::Item(ref mut x) => {
             test(&mut x); //~ ERROR cannot borrow `x` as mutable, as it is not declared as mutable
+            //~| HELP try removing `&mut` here
         }
     }
 }

--- a/src/test/ui/did_you_mean/issue-37139.stderr
+++ b/src/test/ui/did_you_mean/issue-37139.stderr
@@ -5,7 +5,7 @@ LL |             test(&mut x);
    |                  ^^^^^^
    |                  |
    |                  cannot borrow as mutable
-   |                  try removing `&mut` here
+   |                  help: try removing `&mut` here
 
 error: aborting due to previous error
 

--- a/src/test/ui/nll/issue-51191.rs
+++ b/src/test/ui/nll/issue-51191.rs
@@ -1,13 +1,15 @@
+#![allow(unconditional_recursion)]
+
 struct Struct;
 
 impl Struct {
     fn bar(self: &mut Self) {
-        //~^ WARN function cannot return without recursing
         (&mut self).bar();
         //~^ ERROR cannot borrow `self` as mutable, as it is not declared as mutable [E0596]
+        //~^^ HELP try removing `&mut` here
     }
 
-    fn imm(self) {
+    fn imm(self) { //~ HELP consider changing this to be mutable
         (&mut self).bar();
         //~^ ERROR cannot borrow `self` as mutable, as it is not declared as mutable [E0596]
     }
@@ -25,7 +27,8 @@ impl Struct {
     fn mtblref(&mut self) {
         (&mut self).bar();
         //~^ ERROR cannot borrow `self` as mutable, as it is not declared as mutable [E0596]
+        //~^^ HELP try removing `&mut` here
     }
 }
 
-fn main () {}
+fn main() {}

--- a/src/test/ui/nll/issue-51191.rs
+++ b/src/test/ui/nll/issue-51191.rs
@@ -1,9 +1,9 @@
-#![allow(unconditional_recursion)]
-
 struct Struct;
 
 impl Struct {
     fn bar(self: &mut Self) {
+        //~^ WARN function cannot return without recursing
+        //~^^ HELP a `loop` may express intention better if this is on purpose
         (&mut self).bar();
         //~^ ERROR cannot borrow `self` as mutable, as it is not declared as mutable [E0596]
         //~^^ HELP try removing `&mut` here

--- a/src/test/ui/nll/issue-51191.stderr
+++ b/src/test/ui/nll/issue-51191.stderr
@@ -1,3 +1,15 @@
+warning: function cannot return without recursing
+  --> $DIR/issue-51191.rs:4:5
+   |
+LL |     fn bar(self: &mut Self) {
+   |     ^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
+...
+LL |         (&mut self).bar();
+   |         ----------------- recursive call site
+   |
+   = note: `#[warn(unconditional_recursion)]` on by default
+   = help: a `loop` may express intention better if this is on purpose
+
 error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
   --> $DIR/issue-51191.rs:7:9
    |
@@ -36,6 +48,6 @@ LL |         (&mut self).bar();
    |         cannot borrow as mutable
    |         help: try removing `&mut` here
 
-error: aborting due to 5 previous errors
+error: aborting due to 5 previous errors; 1 warning emitted
 
 For more information about this error, try `rustc --explain E0596`.

--- a/src/test/ui/nll/issue-51191.stderr
+++ b/src/test/ui/nll/issue-51191.stderr
@@ -1,26 +1,14 @@
-warning: function cannot return without recursing
-  --> $DIR/issue-51191.rs:4:5
-   |
-LL |     fn bar(self: &mut Self) {
-   |     ^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
-LL |
-LL |         (&mut self).bar();
-   |         ----------------- recursive call site
-   |
-   = note: `#[warn(unconditional_recursion)]` on by default
-   = help: a `loop` may express intention better if this is on purpose
-
 error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
-  --> $DIR/issue-51191.rs:6:9
+  --> $DIR/issue-51191.rs:7:9
    |
 LL |         (&mut self).bar();
    |         ^^^^^^^^^^^
    |         |
    |         cannot borrow as mutable
-   |         try removing `&mut` here
+   |         help: try removing `&mut` here
 
 error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
-  --> $DIR/issue-51191.rs:11:9
+  --> $DIR/issue-51191.rs:13:9
    |
 LL |     fn imm(self) {
    |            ---- help: consider changing this to be mutable: `mut self`
@@ -28,26 +16,26 @@ LL |         (&mut self).bar();
    |         ^^^^^^^^^^^ cannot borrow as mutable
 
 error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
-  --> $DIR/issue-51191.rs:20:9
+  --> $DIR/issue-51191.rs:22:9
    |
 LL |         (&mut self).bar();
    |         ^^^^^^^^^^^ cannot borrow as mutable
 
 error[E0596]: cannot borrow data in a `&` reference as mutable
-  --> $DIR/issue-51191.rs:20:9
+  --> $DIR/issue-51191.rs:22:9
    |
 LL |         (&mut self).bar();
    |         ^^^^^^^^^^^ cannot borrow as mutable
 
 error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
-  --> $DIR/issue-51191.rs:26:9
+  --> $DIR/issue-51191.rs:28:9
    |
 LL |         (&mut self).bar();
    |         ^^^^^^^^^^^
    |         |
    |         cannot borrow as mutable
-   |         try removing `&mut` here
+   |         help: try removing `&mut` here
 
-error: aborting due to 5 previous errors; 1 warning emitted
+error: aborting due to 5 previous errors
 
 For more information about this error, try `rustc --explain E0596`.

